File: src/workspace/workspace.ts
import { lstat, realpath, stat } from "node:fs/promises";
import path from "node:path";

export class WorkspacePathError extends Error {
  public constructor(message: string) {
    super(message);
    this.name = "WorkspacePathError";
  }
}

export class Workspace {
  readonly #root: string;

  private constructor(root: string) {
    this.#root = root;
  }

  public static async open(root: string): Promise<Workspace> {
    if (root.trim() === "") {
      throw new WorkspacePathError("workspace root is empty");
    }
    let canonical: string;
    try {
      canonical = await realpath(path.resolve(root));
    } catch (error) {
      throw new WorkspacePathError(`cannot resolve workspace root: ${messageOf(error)}`);
    }
    const information = await stat(canonical);
    if (!information.isDirectory()) {
      throw new WorkspacePathError("workspace root is not a directory");
    }
    return new Workspace(path.normalize(canonical));
  }

  public get root(): string {
    return this.#root;
  }

  public async resolveExisting(input: string): Promise<string> {
    const candidate = this.#lexicalCandidate(input);
    let canonical: string;
    try {
      canonical = await realpath(candidate);
    } catch (error) {
      throw new WorkspacePathError(`cannot resolve path: ${messageOf(error)}`);
    }
    this.#assertInside(canonical);
    return path.normalize(canonical);
  }

  public async resolveForWrite(input: string): Promise<string> {
    const candidate = this.#lexicalCandidate(input);
    if (candidate === this.#root) {
      throw new WorkspacePathError("workspace root cannot be used as a file");
    }

    try {
      await lstat(candidate);
      return await this.resolveExisting(input);
    } catch (error) {
      if (!isFileSystemError(error, "ENOENT")) {
        if (error instanceof WorkspacePathError) {
          throw error;
        }
        throw new WorkspacePathError(`cannot inspect path: ${messageOf(error)}`);
      }
    }

    let ancestor = path.dirname(candidate);
    while (true) {
      try {
        await lstat(ancestor);
        break;
      } catch (error) {
        if (!isFileSystemError(error, "ENOENT")) {
          throw new WorkspacePathError(`cannot inspect parent path: ${messageOf(error)}`);
        }
      }
      const parent = path.dirname(ancestor);
      if (parent === ancestor) {
        throw new WorkspacePathError("path escapes the workspace root");
      }
      ancestor = parent;
    }

    let canonicalAncestor: string;
    try {
      canonicalAncestor = await realpath(ancestor);
    } catch (error) {
      throw new WorkspacePathError(`cannot resolve parent path: ${messageOf(error)}`);
    }
    this.#assertInside(canonicalAncestor);
    const remainder = path.relative(ancestor, candidate);
    const resolved = path.join(canonicalAncestor, remainder);
    this.#assertInside(resolved);
    return path.normalize(resolved);
  }

  public relative(absolutePath: string): string {
    return path.relative(this.#root, absolutePath).split(path.sep).join("/") || ".";
  }

  #lexicalCandidate(input: string): string {
    if (input.trim() === "") {
      throw new WorkspacePathError("path is empty");
    }
    if (input.includes("\0")) {
      throw new WorkspacePathError("path contains a NUL byte");
    }
    if (path.isAbsolute(input) || path.win32.isAbsolute(input)) {
      throw new WorkspacePathError("absolute paths are outside the workspace");
    }
    const candidate = path.resolve(this.#root, input);
    this.#assertInside(candidate);
    return candidate;
  }

  #assertInside(candidate: string): void {
    const relative = path.relative(this.#root, path.normalize(candidate));
    if (
      relative === ".." ||
      relative.startsWith(`..${path.sep}`) ||
      path.isAbsolute(relative)
    ) {
      throw new WorkspacePathError("path escapes the workspace root");
    }
  }
}

function isFileSystemError(error: unknown, code: string): boolean {
  return (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    (error as NodeJS.ErrnoException).code === code
  );
}

function messageOf(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: internal/workspace/workspace.go
package workspace

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

var ErrOutsideRoot = errors.New("path escapes the workspace root")

// Workspace resolves model-provided paths against a canonical root.
type Workspace struct {
	root string
}

func New(root string) (*Workspace, error) {
	if strings.TrimSpace(root) == "" {
		return nil, errors.New("workspace root is empty")
	}

	abs, err := filepath.Abs(root)
	if err != nil {
		return nil, fmt.Errorf("make workspace root absolute: %w", err)
	}
	real, err := filepath.EvalSymlinks(abs)
	if err != nil {
		return nil, fmt.Errorf("resolve workspace root: %w", err)
	}
	info, err := os.Stat(real)
	if err != nil {
		return nil, fmt.Errorf("stat workspace root: %w", err)
	}
	if !info.IsDir() {
		return nil, fmt.Errorf("workspace root is not a directory: %s", real)
	}

	return &Workspace{root: filepath.Clean(real)}, nil
}

func (w *Workspace) Root() string {
	return w.root
}

// ResolveExisting resolves a path that must already exist. Symlinks are fully
// evaluated before the workspace boundary is checked.
func (w *Workspace) ResolveExisting(path string) (string, error) {
	candidate, err := w.lexicalCandidate(path)
	if err != nil {
		return "", err
	}
	real, err := filepath.EvalSymlinks(candidate)
	if err != nil {
		return "", err
	}
	if err := w.requireInside(real); err != nil {
		return "", err
	}
	return filepath.Clean(real), nil
}

// ResolveForWrite resolves an existing target or checks the nearest existing
// ancestor of a new target. The check narrows accidental escapes but cannot
// remove filesystem TOCTOU races caused by another concurrent process.
func (w *Workspace) ResolveForWrite(path string) (string, error) {
	candidate, err := w.lexicalCandidate(path)
	if err != nil {
		return "", err
	}
	if candidate == w.root {
		return "", errors.New("workspace root cannot be used as a file")
	}

	if _, err := os.Lstat(candidate); err == nil {
		return w.ResolveExisting(path)
	} else if !errors.Is(err, os.ErrNotExist) {
		return "", err
	}

	ancestor := filepath.Dir(candidate)
	for {
		if _, err := os.Lstat(ancestor); err == nil {
			break
		} else if !errors.Is(err, os.ErrNotExist) {
			return "", err
		}
		parent := filepath.Dir(ancestor)
		if parent == ancestor {
			return "", ErrOutsideRoot
		}
		ancestor = parent
	}

	realAncestor, err := filepath.EvalSymlinks(ancestor)
	if err != nil {
		return "", err
	}
	if err := w.requireInside(realAncestor); err != nil {
		return "", err
	}

	remainder, err := filepath.Rel(ancestor, candidate)
	if err != nil {
		return "", err
	}
	resolved := filepath.Join(realAncestor, remainder)
	if err := w.requireInside(resolved); err != nil {
		return "", err
	}
	return filepath.Clean(resolved), nil
}

func (w *Workspace) Relative(path string) string {
	rel, err := filepath.Rel(w.root, path)
	if err != nil {
		return filepath.ToSlash(path)
	}
	return filepath.ToSlash(rel)
}

func (w *Workspace) lexicalCandidate(path string) (string, error) {
	if strings.TrimSpace(path) == "" {
		return "", errors.New("path is empty")
	}
	if filepath.IsAbs(path) || filepath.VolumeName(path) != "" {
		return "", ErrOutsideRoot
	}
	candidate := filepath.Join(w.root, filepath.Clean(path))
	if err := w.requireInside(candidate); err != nil {
		return "", err
	}
	return candidate, nil
}

func (w *Workspace) requireInside(path string) error {
	rel, err := filepath.Rel(w.root, filepath.Clean(path))
	if err != nil {
		return ErrOutsideRoot
	}
	if rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) || filepath.IsAbs(rel) {
		return ErrOutsideRoot
	}
	return nil
}
